File: sdks/typescript/discovery.mjs
// Copyright 2026 J3nna Technologies, LLC
// SPDX-License-Identifier: Apache-2.0
//
// Discovery — how a Node peer finds others on the mesh. It builds and signs its own presence record
// (carrying its grant), gossips it to seed peers' /gossip endpoints, and receives their presence in return.
// Every received record is verified offline (self-signature, and — under an authority root — its grant), so
// a peer admits only authorized peers. Uses the built-in `fetch` (no deps).

import * as wire from "./wire.mjs";

const b64 = (b) => Buffer.from(b).toString("base64");

// Build this peer's signed PresenceRecord (payload + ed25519 signature over the canonical bytes).
export function buildPresence(ident, grant, endpoint, caps, { heartbeat = null, mcpPath = "/mcp" } = {}) {
  if (heartbeat === null) heartbeat = Math.floor(Date.now() / 1000);
  caps = [...caps];
  const payload = {
    protocol: wire.PROTOCOL,
    id: ident.id,
    public_key: b64(ident.publicKey),
    endpoint,
    mcp_path: mcpPath,
    capabilities: caps,
    heartbeat_unix: heartbeat,
    protocol_major: wire.PROTOCOL_MAJOR,
    grant,
    alg: wire.SIG_ALG,
  };
  const sb = wire.presenceSigningBytes({
    protocol: wire.PROTOCOL, alg: wire.SIG_ALG, id: ident.id, publicKey: ident.publicKey,
    endpoint, mcpPath, capabilities: caps,
    protocolMajor: wire.PROTOCOL_MAJOR, grantId: grant.id, heartbeatUnix: heartbeat,
  });
  return { payload, signature: b64(ident.sign(sb)) };
}

// Verify a presence record's self-signature; with `root` set, also verify its grant binds id↔key and is
// authority-signed (the admission check).
export function verifyRecord(rec, root = null) {
  const p = rec.payload;
  const pub = Buffer.from(p.public_key, "base64");
  const sb = wire.presenceSigningBytes({
    protocol: p.protocol, alg: p.alg || "", id: p.id, publicKey: pub,
    endpoint: p.endpoint, mcpPath: p.mcp_path, capabilities: p.capabilities || [],
    protocolMajor: p.protocol_major || 0, grantId: (p.grant || {}).id || "",
    heartbeatUnix: p.heartbeat_unix,
  });
  if (!wire.verify(pub, Buffer.from(rec.signature, "base64"), sb)) return false;
  if (root === null) return true;
  const g = p.grant;
  if (!g || g.subject !== p.id || !Buffer.from(g.public_key, "base64").equals(pub)) return false;
  const gb = wire.grantSigningBytes({
    alg: g.alg || "", id: g.id, subject: g.subject, publicKey: pub, tier: g.tier,
    scopes: g.scopes || [], issuedAt: g.issued_at, notAfter: g.not_after,
    principal: g.principal || "",
  });
  return wire.verify(root, Buffer.from(g.signature, "base64"), gb);
}

async function gossipOnce(seedBase, myRecord, timeout = 10000) {
  const my = myRecord.payload;
  const env = {
    protocol: wire.PROTOCOL,
    digest: { [my.id]: my.heartbeat_unix },
    records: [myRecord],
  };
  const ctrl = new AbortController();
  const t = setTimeout(() => ctrl.abort(), timeout);
  try {
    const r = await fetch(seedBase.replace(/\/+$/, "") + "/gossip", {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(env),
      signal: ctrl.signal,
    });
    if (!r.ok) throw new Error(`gossip -> ${r.status}`);
    const j = await r.json();
    return j.records || [];
  } finally {
    clearTimeout(t);
  }
}

export class Peer {
  constructor(id, mcp, caps) {
    this.id = id;
    this.mcp = mcp; // the peer's reachable MCP URL (endpoint + mcp_path)
    this.caps = caps;
  }

  toString() {
    return `Peer(id=${this.id.slice(0, 8)}…, caps=${JSON.stringify(this.caps)}, mcp=${this.mcp})`;
  }
}

// Gossip our presence to each seed and return the verified peers learned (excluding self), optionally
// filtered to those advertising `wantCap`.
export async function discover(seeds, myRecord, { root = null, wantCap = null } = {}) {
  const myId = myRecord.payload.id;
  const peers = new Map();
  for (const seed of seeds) {
    let records;
    try {
      records = await gossipOnce(seed, myRecord);
    } catch {
      continue;
    }
    for (const rec of records) {
      const p = rec.payload;
      if (p.id === myId || !verifyRecord(rec, root)) continue;
      const caps = p.capabilities || [];
      if (wantCap && !caps.includes(wantCap)) continue;
      peers.set(p.id, new Peer(p.id, p.endpoint.replace(/\/+$/, "") + p.mcp_path, caps));
    }
  }
  return [...peers.values()];
}

File: sdks/rust/src/discovery.rs
// Copyright 2026 J3nna Technologies, LLC
// SPDX-License-Identifier: Apache-2.0
//! Discovery — how a Rust peer finds others on the mesh. It builds and signs its own presence record
//! (carrying its grant), gossips it to seed peers' /gossip endpoints, and receives their presence in return.
//! Every received record is verified offline (self-signature, and — under an authority root — its grant), so
//! a peer admits only authorized peers.

use std::time::{SystemTime, UNIX_EPOCH};

use base64::Engine;
use serde_json::{json, Value};

use crate::identity::Identity;
use crate::{http, wire};

const B64: base64::engine::general_purpose::GeneralPurpose = base64::engine::general_purpose::STANDARD;

fn now_unix() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs())
        .unwrap_or(0)
}

fn caps_from(v: &Value) -> Vec<String> {
    v.as_array()
        .map(|a| {
            a.iter()
                .filter_map(|c| c.as_str().map(|s| s.to_string()))
                .collect()
        })
        .unwrap_or_default()
}

/// Build this peer's signed PresenceRecord (payload + ed25519 signature over the canonical bytes).
/// `grant` is the grant JSON object returned by enrollment; `heartbeat` defaults to now (seconds) when None.
pub fn build_presence(
    ident: &Identity,
    grant: &Value,
    endpoint: &str,
    caps: &[String],
    heartbeat: Option<u64>,
    mcp_path: &str,
) -> Value {
    let heartbeat = heartbeat.unwrap_or_else(now_unix);
    let grant_id = grant.get("id").and_then(|x| x.as_str()).unwrap_or_default();

    let payload = json!({
        "protocol": wire::PROTOCOL,
        "id": ident.id,
        "public_key": B64.encode(ident.public_key),
        "endpoint": endpoint,
        "mcp_path": mcp_path,
        "capabilities": caps,
        "heartbeat_unix": heartbeat,
        "protocol_major": wire::PROTOCOL_MAJOR,
        "grant": grant,
        "alg": wire::SIG_ALG,
    });

    let sb = wire::presence_signing_bytes(
        wire::PROTOCOL,
        wire::SIG_ALG,
        &ident.id,
        &ident.public_key,
        endpoint,
        mcp_path,
        caps,
        wire::PROTOCOL_MAJOR,
        grant_id,
        heartbeat,
    );
    json!({
        "payload": payload,
        "signature": B64.encode(ident.sign(&sb)),
    })
}

/// Verify a presence record's self-signature; with `root` set, also verify its grant binds id↔key and is
/// authority-signed (the admission check).
pub fn verify_record(rec: &Value, root: Option<&[u8]>) -> bool {
    let p = match rec.get("payload") {
        Some(p) => p,
        None => return false,
    };
    let pub_b64 = match p.get("public_key").and_then(|x| x.as_str()) {
        Some(s) => s,
        None => return false,
    };
    let pubkey = match B64.decode(pub_b64.as_bytes()) {
        Ok(b) => b,
        Err(_) => return false,
    };

    let grant_id = p
        .get("grant")
        .and_then(|g| g.get("id"))
        .and_then(|x| x.as_str())
        .unwrap_or_default();

    let sb = wire::presence_signing_bytes(
        p.get("protocol").and_then(|x| x.as_str()).unwrap_or_default(),
        p.get("alg").and_then(|x| x.as_str()).unwrap_or_default(),
        p.get("id").and_then(|x| x.as_str()).unwrap_or_default(),
        &pubkey,
        p.get("endpoint").and_then(|x| x.as_str()).unwrap_or_default(),
        p.get("mcp_path").and_then(|x| x.as_str()).unwrap_or_default(),
        &caps_from(p.get("capabilities").unwrap_or(&Value::Null)),
        p.get("protocol_major").and_then(|x| x.as_u64()).unwrap_or(0) as u32,
        grant_id,
        p.get("heartbeat_unix").and_then(|x| x.as_u64()).unwrap_or(0),
    );

    let sig = match rec
        .get("signature")
        .and_then(|x| x.as_str())
        .and_then(|s| B64.decode(s.as_bytes()).ok())
    {
        Some(s) => s,
        None => return false,
    };
    if !wire::verify(&pubkey, &sig, &sb) {
        return false;
    }

    let root = match root {
        None => return true,
        Some(r) => r,
    };

    let g = match p.get("grant") {
        Some(g) if g.is_object() => g,
        _ => return false,
    };
    let subject = g.get("subject").and_then(|x| x.as_str()).unwrap_or_default();
    let id = p.get("id").and_then(|x| x.as_str()).unwrap_or_default();
    let g_pub = g
        .get("public_key")
        .and_then(|x| x.as_str())
        .and_then(|s| B64.decode(s.as_bytes()).ok())
        .unwrap_or_default();
    if subject != id || g_pub != pubkey {
        return false;
    }

    let scopes: Vec<String> = g
        .get("scopes")
        .map(caps_from)
        .unwrap_or_default();
    let gb = wire::grant_signing_bytes(
        g.get("alg").and_then(|x| x.as_str()).unwrap_or_default(),
        g.get("id").and_then(|x| x.as_str()).unwrap_or_default(),
        subject,
        &pubkey,
        g.get("tier").and_then(|x| x.as_u64()).unwrap_or(0),
        &scopes,
        g.get("issued_at").and_then(|x| x.as_u64()).unwrap_or(0),
        g.get("not_after").and_then(|x| x.as_u64()).unwrap_or(0),
        g.get("principal").and_then(|x| x.as_str()).unwrap_or_default(),
    );
    let g_sig = match g
        .get("signature")
        .and_then(|x| x.as_str())
        .and_then(|s| B64.decode(s.as_bytes()).ok())
    {
        Some(s) => s,
        None => return false,
    };
    wire::verify(root, &g_sig, &gb)
}

fn gossip_once(seed_base: &str, my_record: &Value) -> crate::error::Result<Vec<Value>> {
    let my = &my_record["payload"];
    let id = my.get("id").and_then(|x| x.as_str()).unwrap_or_default();
    let heartbeat = my.get("heartbeat_unix").cloned().unwrap_or(json!(0));
    let env = json!({
        "protocol": wire::PROTOCOL,
        "digest": { id: heartbeat },
        "records": [my_record],
    });
    let url = format!("{}/gossip", seed_base.trim_end_matches('/'));
    let resp = http::post_json(&url, &env, 10.0)?;
    Ok(resp
        .get("records")
        .and_then(|r| r.as_array())
        .cloned()
        .unwrap_or_default())
}

/// A discovered peer on the mesh.
#[derive(Clone, Debug)]
pub struct Peer {
    pub id: String,
    /// The peer's reachable MCP URL (endpoint + mcp_path).
    pub mcp: String,
    pub caps: Vec<String>,
}

/// Gossip our presence to each seed and return the verified peers learned (excluding self), optionally
/// filtered to those advertising `want_cap`.
pub fn discover(
    seeds: &[String],
    my_record: &Value,
    root: Option<&[u8]>,
    want_cap: Option<&str>,
) -> Vec<Peer> {
    let my_id = my_record["payload"]
        .get("id")
        .and_then(|x| x.as_str())
        .unwrap_or_default()
        .to_string();
    let mut peers: std::collections::BTreeMap<String, Peer> = std::collections::BTreeMap::new();

    for seed in seeds {
        let records = match gossip_once(seed, my_record) {
            Ok(r) => r,
            Err(_) => continue,
        };
        for rec in &records {
            let p = match rec.get("payload") {
                Some(p) => p,
                None => continue,
            };
            let pid = p.get("id").and_then(|x| x.as_str()).unwrap_or_default();
            if pid == my_id || !verify_record(rec, root) {
                continue;
            }
            let caps = caps_from(p.get("capabilities").unwrap_or(&Value::Null));
            if let Some(wc) = want_cap {
                if !caps.iter().any(|c| c == wc) {
                    continue;
                }
            }
            let endpoint = p.get("endpoint").and_then(|x| x.as_str()).unwrap_or_default();
            let mcp_path = p.get("mcp_path").and_then(|x| x.as_str()).unwrap_or_default();
            let mcp = format!("{}{}", endpoint.trim_end_matches('/'), mcp_path);
            peers.insert(
                pid.to_string(),
                Peer {
                    id: pid.to_string(),
                    mcp,
                    caps,
                },
            );
        }
    }
    peers.into_values().collect()
}
